src/crimson/osd/object_context.h

```diff
@@ -130,6 +130,15 @@ class ObjectContext : public ceph::common::intrusive_lru_base<
     });
   }
 
+  template <TriMutexSyncLock Lock, typename Func>
+  auto _with_promoted_lock(Lock &lock, Func &&func) {
+    Ref obc = this;
+    lock.lock();
+    return seastar::futurize_invoke(func).finally([&lock, obc] {
+      lock.unlock();
+    });
+  }
```

Comment:
could put something like
```c++
template <typename Mutex, typename Func>
auto _with_lock(Mutex& mutex, Func&& func) {
    Ref obc{this};
    const std::lock_guard<Mutex> lock{mutex};
    return seastar::futurize_invoke(std::forward(func));
}
```

IMHO, no need to use the concept for enforcing the type constraints. because `future<>` cannot be discarded with `[[nodiscard]]` attribute. i think SFINAE should work here, so no need to use a new name, unless we need to make it more visible. but for what reason?

Reply:
Seems not working directly:
```
/root/ceph/src/crimson/osd/object_context.h:138:15: error: ‘co_await’ cannot be used in a function with a deduced return type
  138 |     co_return co_await seastar::futurize_invoke(func);
      |               ^~~~~~~~
/root/ceph/src/crimson/osd/object_context.h:135:8: error: unable to find the promise type for this coroutine
  135 |   auto _with_promoted_lock(Mutex& mutex, Func&& func) {
```

Will fallback to the original version.

Reply:
thanks. will try to come up with a followup.